src/esp32/single_encoder.rs

```diff
@@ -0,0 +1,341 @@
+                    )
+                })?;
+            }
+            Direction::Backwards | Direction::StoppedBackwards => {
+                esp!(unsafe {
+                    esp_idf_sys::pcnt_isr_handler_add(
+                        self.config.unit,
+                        Some(Self::irq_handler_decrement),
+                        self.pulse_counter.as_mut() as *mut PulseStorage as *mut _,
+                    )
+                })?;
+            }
+        };
+
+        unsafe {
+            match esp_idf_sys::pcnt_set_filter_value(self.config.unit, MAX_GLITCH_MICROSEC * 80) {
+                ESP_OK => {}
+                err => return Err(EspError::from(err).unwrap().into()),
+            }
+            match esp_idf_sys::pcnt_filter_enable(self.config.unit) {
+                ESP_OK => {}
+                err => return Err(EspError::from(err).unwrap().into()),
+            }
+        }
+
+        unsafe {
+            match esp_idf_sys::pcnt_event_enable(self.config.unit, pcnt_evt_h_lim) {
+                ESP_OK => {}
+                err => return Err(EspError::from(err).unwrap().into()),
+            }
+            match esp_idf_sys::pcnt_event_enable(self.config.unit, pcnt_evt_l_lim) {
+                ESP_OK => {}
+                err => return Err(EspError::from(err).unwrap().into()),
+            }
+        }
+
+        Ok(())
+    }
+
+    #[inline(always)]
+    #[link_section = ".iram1.pcnt_srv"]
+    unsafe extern "C" fn irq_handler_increment(arg: *mut core::ffi::c_void) {
+        let arg: &mut PulseStorage = &mut *(arg as *mut _);
+        let mut status = 0;
+        esp_idf_sys::pcnt_get_event_status(arg.unit, &mut status as *mut c_ulong);
+        if status & pcnt_evt_h_lim != 0 {
+            arg.acc.fetch_add(1, Ordering::SeqCst);
+        }
+        if status & pcnt_evt_l_lim != 0 {
+            arg.acc.fetch_add(1, Ordering::SeqCst);
+        }
+    }
+
+    #[inline(always)]
+    #[link_section = ".iram1.pcnt_srv"]
+    unsafe extern "C" fn irq_handler_decrement(arg: *mut core::ffi::c_void) {
+        let arg: &mut PulseStorage = &mut *(arg as *mut _);
+        let mut status = 0;
+        esp_idf_sys::pcnt_get_event_status(arg.unit, &mut status as *mut c_ulong);
+        if status & pcnt_evt_h_lim != 0 {
+            arg.acc.fetch_sub(1, Ordering::SeqCst);
+        }
+        if status & pcnt_evt_l_lim != 0 {
+            arg.acc.fetch_sub(1, Ordering::SeqCst);
+        }
+    }
+}
+
+impl Encoder for Esp32SingleEncoder {
+    fn get_properties(&mut self) -> EncoderSupportedRepresentations {
+        EncoderSupportedRepresentations {
+            ticks_count_supported: true,
+            angle_degrees_supported: false,
+        }
+    }
+    fn get_position(&self, position_type: EncoderPositionType) -> anyhow::Result<EncoderPosition> {
+        match position_type {
+            EncoderPositionType::TICKS | EncoderPositionType::UNSPECIFIED => {
```

Comment:
this still feels weird to me. Do we expect an encoder type to support both absolute or Ticks?

Reply:
It feels weird to me too, but currently we assume that as a possibility in our proto API design for encoder, so I felt I had to support it